File: top_censoring/outdetect/02_compare_lis_vs_outdetect.py
"""Compare the LIS Tukey-fence screening (existing method) against the
new outdetect-style (median + MAD, log space, alpha=3, top-tail only)
screening, over the same ~1825 survey-year universe from GMD_all_2017.dta.

Outputs:
  - lis_vs_outdetect_comparison.csv : full universe, both methods side by side,
    with flags for which method(s) singled out each survey as problematic.
  - always_problematic_surveys.csv : the subset flagged by BOTH methods
    (the direct answer to "which surveys are problematic no matter the
    outlier-detection method used").

Flagging rule (documented assumption, see README.md):
  - LIS-flagged  = survey is one of the 24 rows in
    24_surveys_four_indicators_lis_comparison.csv (the team's existing
    ground-truth "problematic" list).
  - outdetect-flagged = survey is among the top 24 surveys (by the same
    metric, "welfare share flagged/above ceiling (%)") under the outdetect
    method, i.e. the same list *size* as the LIS list, so the two methods
    are compared on equal footing rather than an arbitrarily different cutoff.
"""

from pathlib import Path

import pandas as pd

BASE = Path(__file__).resolve().parents[1]  # top_censoring/
OUT_DIR = Path(__file__).resolve().parent / "outputs"

LIS_FULL = BASE / "outputs" / "all_surveys_four_indicators_lis_comparison.csv"
LIS_TOP24 = BASE / "outputs" / "24_surveys_four_indicators_lis_comparison.csv"
OUTDETECT_FULL = OUT_DIR / "outdetect_all_surveys_summary.csv"
WELFARE_TYPE = (
    Path(__file__).resolve().parents[2]
    / "corrected_method"
    / "01-processing"
    / "raw"
    / "interpolated_means.csv"
)

N_TOP = 24  # matches the size of the existing ground-truth LIS list

lis_full = pd.read_csv(LIS_FULL)
lis_top24 = pd.read_csv(LIS_TOP24)
outdetect_full = pd.read_csv(OUTDETECT_FULL)

lis_top24_keys = set(zip(lis_top24["Country"], lis_top24["Year"]))

outdetect_top24 = (
    outdetect_full.sort_values("Welfare share flagged (%)", ascending=False)
    .head(N_TOP)
)
outdetect_top24_keys = set(zip(outdetect_top24["Country"], outdetect_top24["Year"]))

merged = lis_full.merge(
    outdetect_full,
    on=["Country", "Year", "Survey"],
    suffixes=(" (LIS)", " (outdetect)"),
    how="outer",
)

merged["Flagged by LIS"] = merged.apply(
    lambda r: (r["Country"], r["Year"]) in lis_top24_keys, axis=1
)
merged["Flagged by outdetect"] = merged.apply(
    lambda r: (r["Country"], r["Year"]) in outdetect_top24_keys, axis=1
)
merged["Flagged by both (always problematic)"] = (
    merged["Flagged by LIS"] & merged["Flagged by outdetect"]
)

welfare_type = pd.read_csv(WELFARE_TYPE, usecols=["country_code", "year", "welfare_type"])
welfare_type = welfare_type.drop_duplicates(subset=["country_code", "year"])
merged = merged.merge(
    welfare_type,
    left_on=["Country", "Year"],
    right_on=["country_code", "year"],
    how="left",
).drop(columns=["country_code", "year"])

merged = merged.sort_values(
    ["Flagged by both (always problematic)", "Welfare share above ceiling (%)"],
    ascending=[False, False],
)

OUT_DIR.mkdir(parents=True, exist_ok=True)
comparison_path = OUT_DIR / "lis_vs_outdetect_comparison.csv"
merged.to_csv(comparison_path, index=False)

always_problematic = merged[merged["Flagged by both (always problematic)"]]
always_problematic_path = OUT_DIR / "always_problematic_surveys.csv"
always_problematic.to_csv(always_problematic_path, index=False)

print(f"Wrote {len(merged)} rows to {comparison_path}")
print(f"Wrote {len(always_problematic)} rows to {always_problematic_path}")
print()
print("Surveys flagged as problematic by BOTH methods:")
print(
    always_problematic[
        ["Country", "Year", "Survey", "welfare_type", "Welfare share above ceiling (%)", "Welfare share flagged (%)"]
    ].to_string(index=False)
)
print()
print(f"LIS-only count: {(merged['Flagged by LIS'] & ~merged['Flagged by outdetect']).sum()}")
print(f"outdetect-only count: {(~merged['Flagged by LIS'] & merged['Flagged by outdetect']).sum()}")

File: top_censoring/outdetect_r/02_compare_lis_vs_outdetect.R
#!/usr/bin/env Rscript

# Compare the LIS Tukey-fence screening (existing method) against the
# outdetect-EXACT screening (median + S-estimator via robustbase::Sn(),
# log space, alpha=3, top-tail only), over the same ~1,825 survey-year
# universe from GMD_all_2017.dta. Mirrors ../outdetect/02_compare_lis_vs_outdetect.py.
#
# Outputs:
#   - lis_vs_outdetect_comparison.csv : full universe, both methods side by
#     side, with flags for which method(s) singled out each survey.
#   - always_problematic_surveys.csv : the subset flagged by BOTH methods.

LIS_FULL <- file.path("top_censoring", "outputs", "all_surveys_four_indicators_lis_comparison.csv")
LIS_TOP24 <- file.path("top_censoring", "outputs", "24_surveys_four_indicators_lis_comparison.csv")
OUTDETECT_FULL <- file.path("top_censoring", "outdetect_r", "outputs", "outdetect_all_surveys_summary.csv")
WELFARE_TYPE <- file.path("corrected_method", "01-processing", "raw", "interpolated_means.csv")
OUT_DIR <- file.path("top_censoring", "outdetect_r", "outputs")

lis_full <- read.csv(LIS_FULL, stringsAsFactors = FALSE, check.names = FALSE)
lis_top24 <- read.csv(LIS_TOP24, stringsAsFactors = FALSE, check.names = FALSE)
outdetect_full <- read.csv(OUTDETECT_FULL, stringsAsFactors = FALSE, check.names = FALSE)

lis_top24_key <- paste(lis_top24$Country, lis_top24$Year, lis_top24$Survey, sep = "||")
outdetect_full$`Flagged by outdetect` <- suppressWarnings(as.numeric(outdetect_full[["Welfare share flagged (%)"]])) >= 5

merged <- merge(
  lis_full, outdetect_full,
  by = c("Country", "Year", "Survey"),
  suffixes = c(" (LIS)", " (outdetect)"),
  all = TRUE
)

merged$`Flagged by LIS` <- paste(merged$Country, merged$Year, merged$Survey, sep = "||") %in% lis_top24_key
merged$`Flagged by both (always problematic)` <- merged$`Flagged by LIS` & merged$`Flagged by outdetect`

welfare_type <- read.csv(WELFARE_TYPE, stringsAsFactors = FALSE)[, c("country_code", "year", "welfare_type")]
welfare_type <- welfare_type[!duplicated(welfare_type[, c("country_code", "year")]), ]
merged <- merge(
  merged, welfare_type,
  by.x = c("Country", "Year"), by.y = c("country_code", "year"),
  all.x = TRUE
)

lis_share_col <- grep("^Welfare share above ceiling", names(merged), value = TRUE)[1]
if (is.na(lis_share_col)) stop("Could not find the LIS welfare-share column after merge.")
lis_share <- suppressWarnings(as.numeric(merged[[lis_share_col]]))
merged <- merged[order(-merged$`Flagged by both (always problematic)`, -lis_share), ]

dir.create(OUT_DIR, recursive = TRUE, showWarnings = FALSE)
comparison_path <- file.path(OUT_DIR, "lis_vs_outdetect_comparison.csv")
write.csv(merged, comparison_path, row.names = FALSE)

always_problematic <- merged[merged$`Flagged by both (always problematic)`, ]
always_problematic_path <- file.path(OUT_DIR, "always_problematic_surveys.csv")
write.csv(always_problematic, always_problematic_path, row.names = FALSE)

message("Wrote ", nrow(merged), " rows to ", comparison_path)
message("Wrote ", nrow(always_problematic), " rows to ", always_problematic_path)
message("")
message("Surveys flagged as problematic by BOTH methods:")
print(
  always_problematic[, c(
    "Country", "Year", "Survey", "welfare_type",
    "Welfare share above ceiling (%)", "Welfare share flagged (%)"
  )],
  row.names = FALSE
)
message("")
message("LIS-only count: ", sum(merged$`Flagged by LIS` & !merged$`Flagged by outdetect`))
message("outdetect-only count: ", sum(!merged$`Flagged by LIS` & merged$`Flagged by outdetect`))
